### PR TITLE
feat(Packaging): Add patterns setting and deprecate include/exclude

### PR DESCRIPTION
**Pull request by juanjoDiaz**
Closes: #8093

### CODE REVIEW

**Review by medikoo — CHANGES_REQUESTED** [review]
@juanjoDiaz great thanks for that!  Can you ensure that branch is up to date with `master` and that CI passes(?) I see that packing tests failed (those run through `npm run integration-test-run-package`)

**Comment by juanjoDiaz** [comment]
Rebased on top of master and fix the broken test.

**Review by medikoo — CHANGES_REQUESTED** [review]
@juanjoDiaz great thanks for that! It looks very promising, still see my comments.

I believe we need to narrow down scope of this purely, so it's purely about introduction of `patterns`, and doesn't touch any functionality that processes resolved patterns (this can be in scope of other PR)

File: docs/deprecations.md
Comment (juanjoDiaz) [reply docs/deprecations.md:125]: I can tackle this once we are happy with the implementation 🙂

File: lib/plugins/package/lib/zipService.js
Comment (juanjoDiaz) [reply lib/plugins/package/lib/zipService.js:17]: This `excludeDevDependencies ` method received the `package` config and updated the `exclude` property with the dev dependencies.

* It shouldn't be in `zipService` but in `packageService` (I didn't change it to don't make the review even more complicated.)
* The method is really tightly coupled with the include/exclude format. Now, it simple returns the list of devDependencies' patterns so the calling method can add them to the patterns in any way (future-proffness).

File: lib/plugins/package/lib/packageService.js
Comment (juanjoDiaz) [reply lib/plugins/package/lib/packageService.js:70]: That's essentially what I've done.

There are 3 functions that gather the patterns: `resolveFilePathsAll`, `resolveFilePathsFunction and `resolveFilePathsLayer`.

The three of them do the same:
```js
    return this.getExcludes(funcPackageConfig.exclude, true)
      .then(exclude => {
        const params = { exclude, include: this.getIncludes(funcPackageConfig.include) };
        return params;
      })
      .then(params => this.excludeDevDependencies(params))
      .then(params => this.resolveFilePathsFromPatterns(params));
```

So I've abstracted that logic to a `getPatterns` function which calls `getExcludes` and `getIncludes`.
`getIncludes` does exactly as you suggested.
`getExcludes` only change is that it adds the dependencies from `devDependencies`, instead of doing a separate call to the function in `zipService` that would add it to the `exclude` property as I explained in a different comment.

Does that make sense?
I understand that you want to minimize changes, but I think that the code was getting a little bit spaghetti and the implementation in this PR really simplifies things.

File: package.json
Comment (juanjoDiaz) [reply package.json]: I can do a separate commit or a separate PR if you'd rather that.
I've put `micromatch` back as it was and just added a comment.

File: test/unit/lib/plugins/package/lib/packageService.test.js
Comment (juanjoDiaz) [reply test/unit/lib/plugins/package/lib/packageService.test.js]: With my latest changes, there are no changes in these tests anymore.

**Comment by juanjoDiaz** [comment]
Ok, I think that I've tackled everything.

Things left are:
* Are you happy with my change to the `excludeDevDependencies` method? In short, instead of a function that appends, to the `exclude` setting, a pure function that returns the patterns to append.
* Are you happy with the current `getPatterns` function?

Once we agree on those, and in any other that I've missed, I'll update the docs.

File: lib/plugins/package/lib/zipService.js
Comment (medikoo) [reply lib/plugins/package/lib/zipService.js:17]: Thanks for clarifications. Actually `excludeDevDependencies` doesn't really receive `package` config directly.

The actual packaging config is consumed in two places:
- https://github.com/serverless/serverless/blob/854526816265a425d2d3fe7ae4285b242893b434/lib/plugins/package/lib/packageService.js#L21
- https://github.com/serverless/serverless/blob/854526816265a425d2d3fe7ae4285b242893b434/lib/plugins/package/lib/packageService.js#L197-L201

Then `this.getIncludes` and `this.getExcludes` resolve internal _include_ and _exclude_ patterns collections (which are not `package` config objects) which are processed further.

Technically to implement this functionality, all we need is to tweak `this.getIncludes` so it combines `package.patterns` into returned _include_ patterns, and otherwise nothing else is needed to be done I think.

Of course it'll be also nice to clear internals so they also process single _patterns_ collections and not both _include_ and _exclude_, but as it can be seen a separate concern, and it's best if it's addressed with other PR (ideally PR's should be minimal and focus on one thing, especially that we deal here with really sensitive functionality).

Additionally we have a refactor planned to how we resolve files in a service and dev dependencies: https://github.com/serverless/serverless/issues/8586 and that will replace some of the logic in `zipService` and `packageService`, so it may not be worth to imply any refactors to it at this point

File: package.json
Comment (medikoo) [reply package.json]: > I can do a separate commit or a separate PR if you'd rather that.

Yes, that'll be great, and ideally if such change is first discussed, so before jumping into implementation we have full agreement that given refactor is justified. Please check https://github.com/serverless/serverless/issues/8586 it's probably a good point for discussing this

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:68]: > That's essentially what I've done.

Changes I see now, look as way more than needed to accomplish what's described above, and here we deal with really sensitive functionality, that's why I'm pushing for minimal needed changeset 

I wouldn't touch `getExcludes` and I don't think any changes are needed in `zipService`. I've elaborated on that a bit more here: https://github.com/serverless/serverless/pull/8581#discussion_r539490987

Also, sorry I don't want to sound dismissive, in overall you've done a really great work.

**Comment by juanjoDiaz** [comment]
This is now 3 commits.
* Hyper-minimal change as you suggested
* Use patterns in the create plugin
* Use patterns in the docs

Let me know if I still miss anything.


All the internals are still working with include/exclude so I'd really suggest doing a separate PR once this is merged to consolidate the internals to use just `patterns`

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [inline lib/plugins/package/lib/packageService.js:22]: I think we need to support both `include` and `patterns`, in a way I proposed in this comment: https://github.com/serverless/serverless/pull/8581#discussion_r538467304

Technically we stand between choice:
1. If user provided `patterns`, throw if any `include` and `exclude` configurations are found
2. Support all three

I think it's easier (and not harmful) to support all three

**Review by medikoo — CHANGES_REQUESTED** [review]
@juanjoDiaz It looks great! I've just suggested a more permissive adaptation (so we support all three forms together with deprecation notice).

Also we need some tests, and for that ideally if https://github.com/serverless/serverless/issues/8594 is addressed first. Having that it'll be very easy to add few tests that confirm on `patterns` support

File: docs/guides/monitoring/README.md
Comment (juanjoDiaz) [reply docs/guides/monitoring/README.md:102]: This is a readme example that use to do:
```yaml
package:
  include:
    - src/*.js
    - dist/*.js
    - dist/*.js.map
```

but I think that's a better example to exclude sources if we add the compiled files with their source-maps from `dist` and exclude the original files from `src`.

Do you dissagree?

File: docs/providers/aws/guide/packaging.md
Comment (juanjoDiaz) [reply docs/providers/aws/guide/packaging.md:91]: This is example in `aliyun`, `azure`, `google`, `kubeless`, `openwhisk` and `tencent`.
So I think that it makes sense to use the same structure in `aws` for consistency (I fixed all the docs comparing them side by side).

Do you disagree?

File: lib/plugins/package/lib/packageService.js
Comment (juanjoDiaz) [reply lib/plugins/package/lib/packageService.js:22]: You also suggested to crash if both are used in comment: https://github.com/serverless/serverless/issues/8093#issuecomment-676003542 😅

The problem of supporting both `include` and `patterns` is that the order in which they are appended becomes non-evident to the user and can cause tickets simply because the user expects a different order.

The current implementation support using `include/exclude` at the service level and `patterns` at the function level or viceversa. It errors if you use `include/exclude` and `patterns` in the same package object (done during the initialization validation).

I think that it's safer to avoid user confusion.

Do you still think that we should allow using both at the same time?

File: docs/cn/providers/tencent/components/basic-components/tencent-scf.md
Comment (juanjoDiaz) [reply docs/cn/providers/tencent/components/basic-components/tencent-scf.md:87]: If `@servereless/components` uses `serverless`, this will work just as it used to.

This will need to change for the next major version update anyway.
I think that keeping it is creating future trouble.

Do you disagree?

File: docs/guides/monitoring/README.md
Comment (medikoo) [reply docs/guides/monitoring/README.md:102]: @juanjoDiaz this example intentionally includes sources in a bundle (note the sentence: _To use a sourcemap, ensure that your packaging directory includes the compiled source_)

Weather it's good example, and whether it could be improved it's the matter that's out of the scope of this PR. We may eventually open another issue to discuss it.

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [reply docs/providers/aws/guide/packaging.md:91]: I wouldn't change in scope of this PR (it's not a matter we're addressing here). Also I think AWS version is better.

Also this documentation is scheduled to be secluded to general section and removed from providers, where it's duplicated in multiple places. Check https://github.com/serverless/serverless/issues/8085 & https://github.com/serverless/serverless/issues/8097

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:22]: > You also suggested to crash if both are used in comment: #8093 (comment) 😅

That's true. Sorry for confusion. At that point it was not deeply thought. Thing is that crashing is more obtrusive.

Firstly the easy approach to this is to upfront investigate all `package` configs (service-level, function-level, layer-level) and crash if we find that there's `patterns` somewhere and `includes` or `excludes` elsewhere`, and that's aggressive if e.g. user is switching gradually, e.g. refactor first function one, check how it works and then refactor others. It'll be nice to not break that.

Now trying to validate case by case, is confusing, as e.g. for function 1 config (which uses just `include` and `exclude`) may be valid and for function 2 (which uses just `patterns`) may be not (when on service level `include` or `exclude` is used)

I assume it's not harmful to support all (user will always be approached with deprecation notice anyway), and implementation wise it's simplest approach.

> The problem of supporting both include and patterns is that the order in which they are appended becomes non-evident to the user and can cause tickets simply because the user expects a different order.

User should not reason about order in such case, but simply ensure that just `patterns` is used , also again _deprecation_ notice will signal that config stands on quirky ground

File: docs/cn/providers/tencent/components/basic-components/tencent-scf.md
Comment (medikoo) [reply docs/cn/providers/tencent/components/basic-components/tencent-scf.md:87]: > If @servereless/components uses serverless,

It' doesn't use `serverless`, it's com (package internals), it's a completely different CLI

File: lib/plugins/package/lib/packageService.js
Comment (juanjoDiaz) [reply lib/plugins/package/lib/packageService.js:22]: > Now trying to validate case by case, is confusing, as e.g. for function 1 config (which uses just include and exclude) may be valid and for function 2 (which uses just patterns) may be not (when on service level include or exclude is used)

That's not how it's implemented in this final version of the implementation. Function 1 and 2 will be valid in that example.

Function 3 which uses `patterns` and `exclude` under the same package object would be invalid.
I.e, to get the error you need something like:
```yaml
package:
  patterns:
    - my-file.js
  include:
    - my-other-file.js
```

I think that's reasonable. However, if you still disagree, let me know and I'll just remove it. I'm running out of the time I've allocated for this 😅

Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:22]: > Function 1 and 2 will be valid in that example.

I think if we invalidate it on function level, we should definitely also not support combining `include` and `exclude` on service level with `patterns` on function level. That would be even more confusing (what's the logical reason for rejecting just one case?)

Anyway have you read the arguments I've put for supporting all without throwing (?) I take it's simplest thing we can do, and really not harmful

Comment (juanjoDiaz) [reply lib/plugins/package/lib/packageService.js:22]: I'm not sure if we are understanding each other completely or getting a bit tangled up. 😅
The validation for service level is exactly the same at the function level.
Let's give it one more go and then do whatever change it needed 😄

Currently:
* You can use `patterns` OR `include/exclude` at the service level.
* You can use `patterns` OR `include/exclude` at the function level.
* It doesn't matter if you use pattern in `patterns` at service and `include/exclude` at the function
* It doesn't matter if you use pattern in `include/exclude` at service and `patterns` at the function
* etc...

The only thing that fails, is to mix both `patterns` and `include/exclude` in the same config (regardless of whether it is in the service or function level).

Valid:
```yaml
package:
  patterns:
    - my-file.js

functions:
  myFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Valid:
```yaml
package:
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    patterns:
      - my-file.js
```

Valid:
```yaml
package:
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    patterns:
      - my-file.js
  myOtherFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Throws:
```yaml
package:
  patterns:
    - my-file.js
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```
Throws:
```yaml
package:
  patterns:
    - my-file.js

functions:
  myFunction:
    package:
      patterns:
        - my-file.js
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Just repeating myself, the reason to not allow that merging is that service goes before package regardless of whether it's `patterns` or `include/exclude`. And `exclude` goes before `include`.
However,  if we mix `include/exclude/pattern` in the same config. We need to decide in which order to merge them. 
It's trivial, I know, `exclude -> include -> patterns`, but users might expect otherwise and create confusion.

> User should not reason about order in such case, but simply ensure that just patterns is used , also again deprecation notice will signal that config stands on quirky ground

That's right! That's why I thought that all combinations should be allowed, except for the one that combines both in the same object.

Was this already clear to you?

Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:22]: > You can use patterns OR include/exclude at the function level.
> It doesn't matter if you use pattern in patterns at service and include/exclude at the function

In last comment, my point was that above feels confusing. If we already support mixing both `include` and `patterns` (coming from two configuration sources) to package given lambda, why restrict mixing of `include` and `patterns` coming from one configuration source.

> However, if we mix include/exclude/pattern in the same config. We need to decide in which order to merge them.

Yes, I've proposed order in one of the first comments here: https://github.com/serverless/serverless/pull/8581#discussion_r538467304

> It's trivial, I know, exclude -> include -> patterns, but users might expect otherwise and create confusion.

User should not investigate and debate on such order, instead should just use `pattern`. We'd allow mixing both sas temporary mean simply for ease of migration (no need to be too restrictive here)

**Comment by juanjoDiaz** [comment]
Alright, error removed.
Mixing is now allowed

**Comment by medikoo** [comment]
@juanjoDiaz is this ready for re-review? If it's the case please re-request it (in reviewers box). Thank you!

**Review by medikoo — CHANGES_REQUESTED** [review]
@juanjoDiaz Thanks for update. Can you ensure branch is up to date with master.

Currently merge status show conflicts

**Comment by juanjoDiaz** [comment]
Done :)

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [reply docs/providers/aws/guide/packaging.md:91]: Let's not update this header

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [inline docs/providers/aws/guide/packaging.md:124]: Same here

File: docs/providers/azure/guide/serverless.yml.md
Comment (medikoo) [inline docs/providers/azure/guide/serverless.yml.md:140]: YAML error sneaked in, and let's ensure to reflect original example

File: lib/plugins/aws/provider.js
Comment (medikoo) [inline lib/plugins/aws/provider.js:965]: Let's maintain alphabetical order

File: lib/plugins/package/package.js
Comment (medikoo) [inline lib/plugins/package/package.js:70]: Let's reword it to: _Support for`package.include` and `package.exclude` will be removed with next major release. Please use `package.patterns` instead_

File: lib/plugins/package/package.js
Comment (medikoo) [inline lib/plugins/package/package.js:64]: Let's also confirm on `layers[].package`

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @juanjoDiaz ! We're getting close. Please see my remarks. Also we need to back it with some tests (but with that let's wait for https://github.com/serverless/serverless/issues/8594 which is being addressed by @scadu )

File: lib/plugins/aws/provider.js
Comment (juanjoDiaz) [reply lib/plugins/aws/provider.js:965]: They were not in alphabetical order in the service-level schema. I fixed it there.

**Comment by juanjoDiaz** [comment]
All amended.

> Also we need to back it with some tests (but with that let's wait for #8594 which is being addressed by @scadu )

do you mean, not merging this until #8594 is done?
That might be a PITA regarding merge conflicts and constant rebasing...

File: lib/plugins/aws/provider.js
Comment (medikoo) [reply lib/plugins/aws/provider.js]: > I fixed it there.

Good call, thank you!

**Comment by medikoo** [comment]
> do you mean, not merging this until #8594 is done?

I mean that we definitely should not merge it without tests 

Now, we may hold configuring tests until #8594  is done, or incorporate tests into current spec as here: https://github.com/serverless/serverless/blob/44d398694fc803bb45a07e7bcc0d6cd9e572a704/test/unit/lib/plugins/package/lib/packageService.test.js#L456-L659 (still that is likely to collide with work that @scadu possibly already started)

@scadu what is the current status of #8594 ?

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [inline docs/providers/aws/guide/packaging.md:55]: Let's not change this header

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [reply docs/providers/aws/guide/packaging.md:91]: This seems not addressed

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks for update @juanjoDiaz ! Please see my remarks

**Comment by juanjoDiaz** [comment]
I addressed those last 2.

**Comment by scadu** [comment]
> > do you mean, not merging this until #8594 is done?
> 
> I mean that we definitely should not merge it without tests
> 
> Now, we may hold configuring tests until #8594 is done, or incorporate tests into current spec as here:
> 
> https://github.com/serverless/serverless/blob/44d398694fc803bb45a07e7bcc0d6cd9e572a704/test/unit/lib/plugins/package/lib/packageService.test.js#L456-L659
> 
> (still that is likely to collide with work that @scadu possibly already started)
> @scadu what is the current status of #8594 ?

@medikoo I've barely started, so there should be no clash here. I plan to finish #8594 later this week - maybe even weekend, depending on how much free time I'd have.

**Comment by medikoo** [comment]
>@medikoo I've barely started, so there should be no clash here. I plan to finish #8594 later this week - maybe even weekend, depending on how much free time I'd have.

Thanks for status @scadu !

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [reply docs/providers/aws/guide/packaging.md:91]: Not addressed

File: docs/providers/aws/guide/packaging.md
Comment (medikoo) [inline docs/providers/aws/guide/packaging.md:124]: We should not change this header

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks for update @juanjoDiaz still I think wrong header in doc was updated

**Comment by medikoo** [comment]
@juanjoDiaz I see there were some commits pasted after last review. Is this ready for re-review? If so, can update with master and re-request the review (in reviewers box).

Thank you!

**Comment by juanjoDiaz** [comment]
A far as I remember there was no changes after the latest review.
I've rebased on top of master.

**Review by medikoo — CHANGES_REQUESTED** [review]
@juanjoDiaz great thanks for update. Still currently tests crash, have you seen the CI status?

**Comment by juanjoDiaz** [comment]
Could we get this merged asap?
I keep rebasing but we keep getting merge conflicts because of other PRs taking precedence.

**Comment by medikoo** [comment]
> Could we get this merged asap?
I keep rebasing but we keep getting merge conflicts because of other PRs taking precedence.

@juanjoDiaz Thanks for update. Sure, we'll look into that shortly. Note that it was stale for a long while, as review was not re-requested. I found that it has some new pushes, simply during my maintenance walk through over open PR's.

**Review by medikoo — APPROVED** [review]
@juanjoDiaz Looks great to me!

@pgrzesik I've reviewed this PR back in December, and now just re-reviewed changes after merge update. It all looks good to me, but can you re-review this PR with a fresh eye (please merge if there are no doubts)

File: docs/deprecations.md
Comment (pgrzesik) [inline docs/deprecations.md:22]: Could you please add `Deprecation code:` in a similar way to all other deprecations below?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thanks @juanjoDiaz :bow: I have only two small things - one about adding `deprecation code` to `deprecations.md` section that I've mentioned in other comment. Another thing is that I'd like you to ask to update `serverless.yml.md` reference file: https://github.com/serverless/serverless/blob/master/docs/providers/aws/guide/serverless.yml.md so it reflects changes introduced in this PR. After having these, we're good to go :100:

**Comment by medikoo** [comment]
@juanjoDiaz do you plan to finalize this PR? If you lack resources we totally understand and we will happily take over, just let us know

**Comment by juanjoDiaz** [comment]
Hi @medikoo ,

Can you clarified what's left?

I've been attending all of your requests AFAIK. But then it goes stale and conflicts start arising as master moves forward.
I just rebased on top of master and this should be good to go unless there is anything else.

**Comment by medikoo** [comment]
> Can you clarified what's left?

CI status was red, and we didn't have any notification that's ready for review.

When it's ready for review, please re-request it in top right box. Thank you!

**Comment by juanjoDiaz** [comment]
Alright!
CI is green and I've re-requested review to both reviewers.

**Review by pgrzesik — APPROVED** [review]
Thank you @juanjoDiaz :bow: Looks good - @medikoo, could you also take a look at it before merging?

**Review by medikoo — APPROVED** [review]
Looks great! Thank you @juanjoDiaz

**Comment by tachyon83** [comment]
@juanjoDiaz @medikoo 
I am very sorry to bother you guys. I am having a problem including some files in my serverless deployed function.

I have tried two ways:
1. package.exclude everything then include what I need
2. package.patterns to include what I need

But both are not working. I am just guessing the features you guys developed here are not applicable when using `serverless-bundle` plugin?

```
service: #####

plugins:
  - serverless-dotenv-plugin
  - serverless-bundle
  - serverless-offline

provider:
  name: aws
  runtime: nodejs14.x
  stage: ${opt:stage, 'local'}
  memorySize: 1024
  timeout: 25
  # some private properties are omitted

package:
  individually: true

custom:
  sls:
    version: v1
  serverless-offline:
    httpPort: #####
    lambdaPort: #####
    noPrependStageInUrl: true
    allowCache: true
  dotenv:
    logging: true
  bundle:
    sourcemaps: false
    externals: all

functions:
  api:
    handler: src/main.handler
    name: main-function
    events:
      - httpApi:
          method: any
          path: /
      - httpApi:
          method: any
          path: /{proxy+}

  test:
    handler: src/e2eTest.handler
    name: e2e-test-function
    timeout: 900
    package:
      exclude:
        - ./**
      include:
        - "src/test/**"
```

**Comment by medikoo** [comment]
> But both are not working. I am just guessing the features you guys developed here are not applicable when using serverless-bundle plugin?

It could be the case. Best is to confirm with plugin directly (and I see you've opened https://github.com/AnomalyInnovations/serverless-bundle/issues/363)

**Comment by juanjoDiaz** [comment]
A quick search of `serverless-bundle` codebase shows that they have not updated to `patterns` yet.

**Pull request merged**
